### DISCUSSION TITLE
Aggrid setting rowClassRules

### DISCUSSION THREAD

**Discussion by pingretispec** [zauberzeug/nicegui#2995] [Q&A] [answered]
### Question

I have a ui.aggrid table that is initialized with the following code:
```py
table= ui.aggrid({'defaultColDef': {'flex': 1},
                                             'columnDefs': [{'headerName': 'col1', 'field': 'col1'},
                                                            {'headerName': 'col2', 'field': 'col2'},
                                                            {'headerName': 'col3', 'field': 'col3', 'hide': True},
                                                            {'headerName': 'col4', 'field': 'col4'},
                                                            {'headerName': 'col5', 'field': 'col5', 'hide': True},
                                                            {'headerName': 'col6', 'field': 'col6'},],
                                             'rowData': [],
                                             'rowClassRules': {'bg-red-300': 'data.col3 == "string"'}
})
```
I then update the rowData on a timely basis, which does:
```py
table.options[rowData] = data
table.update()
```

However, I wasn't able to change the background color of any of the rows even if I change the rowClassRules to be `{'bg-red-300': 'true'}`

**Comment by falkoschindler** [comment]
Hi @pingretispec, that's very strange. I reduced it to
```py
ui.aggrid({
    'columnDefs': [
        {'headerName': 'col1', 'field': 'col1'},
        {'headerName': 'col2', 'field': 'col2'},
    ],
    'rowData': [{'col1': 'x', 'col2': 'y'}],
    'rowClass': 'bg-green-300',
})
```
but still can't see any background color. What's going on? Are we using AG Grid incorrectly?

**Reply by thetableman** [reply]
I have managed to get `rowClassRules` working in several tables but in my experience some CSS properties need the `!important` parameter to work. This means that you can't always reference Quasar or Tailwind color classes directly, instead you have to create a custom class that changes the `background-color` property to the desired color with `!important` specified.

Because you can't reference Tailwind/Quasar classes from your own classes easily I define CSS variables that match my most commonly used colors which can then be referenced in these custom classes needed for AG Grid.

```py
from nicegui import ui

ui.add_css('''
    :root {
        --my-bg-green-300-color: #86efac;
    }
           
    .my-bg-green-300-class {
        background-color: var(--my-bg-green-300-color) !important;
    }''')

ui.aggrid({
    'columnDefs': [
        {'headerName': 'col1', 'field': 'col1'},
        {'headerName': 'col2', 'field': 'col2'},
    ],
    'rowData': [{'col1': 'x', 'col2': 'y'}],
    'rowClass': 'my-bg-green-300-class',
})
```

**Comment by falkoschindler** [comment, marked as answer]
Thanks, @thetableman! We need to enforce our background color with the `!important` modifier.

My `rowClass` example works like this:
```py
ui.aggrid({
    'columnDefs': [
        {'headerName': 'col1', 'field': 'col1'},
        {'headerName': 'col2', 'field': 'col2'},
    ],
    'rowData': [{'col1': 'x', 'col2': 'y'}],
    'rowClass': '!bg-green-300',
})
```

And we can use `rowClassRules` like this:
```py
ui.aggrid({
    'columnDefs': [
        {'headerName': 'col1', 'field': 'col1'},
        {'headerName': 'col2', 'field': 'col2'},
    ],
    'rowData': [{'col1': 'x', 'col2': 'y'}],
    'rowClassRules': {':!bg-red-300': '(params) => params.data.col2 == "y"'},
})
```

The colon ":" indicates an attribute that needs to be converted into a JavaScript function on the client. And its signature can be found in the [AG Grid documentation](https://www.ag-grid.com/javascript-data-grid/row-styles/#row-class-rules).

**Reply by thetableman** [reply]
Thanks @falkoschindler! I didn't know you could prefix '!' to a Tailwind class to mark as `!important`, that's a neat little trick.

**Reply by pingretispec** [reply]
Thanks @falkoschindler! Adding the ":!" got it working. It's also interesting that it only works with '(params) => params.data.col2 == "y"' and not 'data.col2 == "y"', which is what aggrid have in their example:
```py
const gridOptions = {
    rowClassRules: {
        'rag-green': 'data.age < 20',
        'rag-amber': 'data.age >= 20 && data.age < 25',
        'rag-red': 'data.age >= 25',
    }}
```

**Reply by falkoschindler** [reply]
Ah, when using the notation without `params`, you don't need the colon:
```py
ui.aggrid({
    'columnDefs': [
        {'headerName': 'col1', 'field': 'col1'},
        {'headerName': 'col2', 'field': 'col2'},
    ],
    'rowData': [{'col1': 'x', 'col2': 'y'}],
    'rowClassRules': {'!bg-red-300': 'data.col2 == "y"'},
})
```

This is because `data.col2 == "y"` can't be converted into a JavaScript function, but needs to be parsed by AG Grid in some other way. The colon is only used for sending JavaScript functions from the Python world, where we need to define them as strings. But if AG Grid expects a string anyway, we don't need the conversion.
